add int8 matmul support to CUDA backend

verstatx opened the pull request:
Description
-----------
Adds support for int8 matmul in the CUDA backend using cublasGemmEx functions. This modifies the gemm functions' api to support a different output array type, so all backends were modified.

This PR depends on s8 support: #3507 
Fixes: #1656 

Checklist
---------
<!-- Check if done or not applicable -->
- [x] Rebased on latest master with #3507
- [x] Code compiles
- [x] Tests pass
- [x] Functions documented

edwinsolisf reviewed (approved):
Tested on RTX 3070Ti all backends on Ubuntu & Windows

verstatx commented:
Squashed commits and rebased on master. Included a minor lint fix to whitespace that I apparently missed. My apologies if this interfered with your process!
